RealmSwift/SyncSubscription.swift

```diff
@@ -220,235 +316,597 @@ import Realm.Private
+    private let rlmSubscriptionSet: RLMSyncSubscriptionSet
+    private let realm: Realm
+    private var index: Int = -1
+
+    init(_ rlmSubscriptionSet: RLMSyncSubscriptionSet, _ realm: Realm) {
+        self.rlmSubscriptionSet = rlmSubscriptionSet
+        self.realm = realm
+    }
+
+    private func nextIndex(for index: Int?) -> Int? {
+        if let index = index, index < self.rlmSubscriptionSet.count - 1 {
+            return index + 1
         }
+        return nil
     }
 
+    mutating public func next() -> AnyQueryResults? {
+        if let index = self.nextIndex(for: self.index) {
+            self.index = index
+            return rlmSubscriptionSet.object(at: UInt(index)).map {
+                AnyQueryResults($0, realm.dynamicObjects($0.objectClassName))
+            }
+        }
+        return nil
+    }
+}
+
+#if swift(>=5.6) && canImport(_Concurrency)
+@available(macOS 10.15, tvOS 13.0, iOS 13.0, watchOS 6.0, *)
+extension SyncSubscriptionSet {
     /**
-     Removes a subscription with the specified query.
+     Asynchronously creates and commit a write transaction and updates the subscription set,
+     this will not wait for the server to acknowledge and see all the data associated with this
+     collection of subscription.
 
-     - warning: This method may only be called during a write subscription block.
+     - parameter block: The block containing the subscriptions transactions to perform.
 
-     - parameter type: The type of the object to be removed.
-     - parameter to: A query for the subscription to be removed from the subscription set.
+     - throws: An `NSError` if the transaction could not be completed successfully.
+               If `block` throws, the function throws the propagated `ErrorType` instead.
      */
-    public func remove<T: Object>(ofType type: T.Type, _ query: @escaping (Query<T>) -> Query<Bool>) {
-        rlmSyncSubscriptionSet.removeSubscription(withClassName: "\(T.self)",
-                                                  predicate: query(Query()).predicate)
+    @MainActor
+    private func write(_ block: (() -> Void)) async throws {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+            write(block) { error in
+                if let error = error {
+                    continuation.resume(throwing: error)
+                } else {
+                    continuation.resume()
+                }
+            }
+        }
+    }
+
+    @MainActor
+    private func subscribe(_ subscriptions: _QuerySubscription...) async throws {
+        try await write {
+            subscriptions.forEach { subscription in
+                self.append(subscription)
+            }
+        }
     }
 
     /**
-     Removes a subscription with the specified query.
+     Appends the query to the current subscription set and wait for the server to acknowledge the subscription,
```

Comment:
Because we are extending from `RealmCollectionImpl` on QueryResults to make use of the default implementation of a realm  ollection, we need to this to be optional.